xbmc/platform/android/activity/JNIXBMCRecognitionListener.cpp

```diff
@@ -0,0 +1,94 @@
+
+#include "JNIXBMCRecognitionListener.h"
+
+#include "CompileInfo.h"
+#include "ServiceBroker.h"
+#include "dialogs/GUIDialogKaiToast.h"
+#include "guilib/GUIComponent.h"
+#include "guilib/GUIWindow.h"
+#include "guilib/GUIWindowManager.h"
+#include "guilib/LocalizeStrings.h"
+#include "messaging/ApplicationMessenger.h"
+#include "utils/log.h"
+
+#include <androidjni/Context.h>
+#include <androidjni/jutils-details.hpp>
+
+using namespace jni;
+
+static std::string s_className =
+    std::string(CCompileInfo::GetClass()) + "/interfaces/XBMCRecognitionListener";
+
+CJNIXBMCRecognitionListener::CJNIXBMCRecognitionListener() : CJNIBase(s_className)
+{
+  m_object = new_object(CJNIContext::getClassLoader().loadClass(GetDotClassName(s_className)));
+  m_object.setGlobal();
+
+  add_instance(m_object, this);
+}
+
+void CJNIXBMCRecognitionListener::RegisterNatives(JNIEnv* env)
+{
+  jclass cClass = env->FindClass(s_className.c_str());
+  if (cClass)
+  {
+    JNINativeMethod methods[] = {
+        {"_onReadyForSpeech", "(Landroid/os/Bundle;)V",
+         reinterpret_cast<void*>(&CJNIXBMCRecognitionListener::_onReadyForSpeech)},
+        {"_onError", "(I)V", reinterpret_cast<void*>(&CJNIXBMCRecognitionListener::_onError)},
+        {"_onResults", "(Landroid/os/Bundle;)V",
+         reinterpret_cast<void*>(&CJNIXBMCRecognitionListener::_onResults)},
+    };
+
+    env->RegisterNatives(cClass, methods, sizeof(methods) / sizeof(methods[0]));
+  }
+}
+
+CJNIXBMCRecognitionListener::~CJNIXBMCRecognitionListener()
+{
+  remove_instance(this);
+}
+
+void CJNIXBMCRecognitionListener::_onReadyForSpeech(JNIEnv* env, jobject thiz, jobject bundle)
+{
+  CGUIDialogKaiToast::QueueNotification(CGUIDialogKaiToast::Info, g_localizeStrings.Get(39177),
+      g_localizeStrings.Get(39179));
+
+  CLog::Log(LOGDEBUG, "CJNIXBMCRecognitionListener onReadyForSpeech()");
+}
+
+void CJNIXBMCRecognitionListener::_onError(JNIEnv* env, jobject thiz, jint i)
+{
+  if (i == CJNISpeechRecognizer::ERROR_NO_MATCH)
+    CGUIDialogKaiToast::QueueNotification(CGUIDialogKaiToast::Info, g_localizeStrings.Get(39177),
+        g_localizeStrings.Get(39180));
+  else
+    CGUIDialogKaiToast::QueueNotification(CGUIDialogKaiToast::Info, g_localizeStrings.Get(39177),
+        g_localizeStrings.Get(39181));
+
+  CLog::Log(LOGDEBUG, "CJNIXBMCRecognitionListener onError(): [{}]", i);
+}
+
+void CJNIXBMCRecognitionListener::_onResults(JNIEnv* env, jobject thiz, jobject bundle)
+{
+  CJNIBundle b = CJNIBundle(jhobject::fromJNI(bundle));
+
+  CJNIArrayList results = b.getStringArrayList(CJNISpeechRecognizer::RESULTS_RECOGNITION);
+
+  const std::string result = results.get(0);
+
+  CGUIWindow* window = CServiceBroker::GetGUI()->GetWindowManager().GetWindow(
+      CServiceBroker::GetGUI()->GetWindowManager().GetActiveWindowOrDialog());
+  CGUIMessage msg(GUI_MSG_SET_TEXT, 0, window->GetFocusedControlID());
+  msg.SetLabel(result);
+  CServiceBroker::GetAppMessenger()->SendGUIMessage(msg, window->GetID());
```

Comment:
Same approach as for ::_onError. define a new message TMSG_SPEECH_RECOGNIZER_RESULTS, disapach via AppMessenger to Capp, then to CGUIDialogKeyboardGeneric::OnSpeechRecognizerResult, which does the SET_TEXT.